src/hooks/useDebouncedValue.ts

```diff
@@ -0,0 +1,17 @@
+import { useEffect, useState } from "react";
+
+function useDebouncedValue<T = unknown>(value: T, timeout: number) {
+  const [updatedValue, setUpdatedValue] = useState<T | null>(null);
```

Comment:
It might be worth to use `lodash.debounce` here to be super sure of the behaviour. We are already using it in `useDebounce`

Reply:
Not sure how that's supposed to work with debounce to be fair... I changed the implementation to this:
```
function useDebouncedValue<T = unknown>(value: T, timeout: number) {
  const [updatedValue, setUpdatedValue] = useState<T | null>(null);
  useEffect(() => {
    debounce(() => setUpdatedValue(value), timeout);
  }, [value, timeout]);

  return updatedValue;
}
```

But my tests are failing.

Reply:
Also just wondering... do we really need `lodash` here? I mean, this hook is pretty straightforward and I think that `setTimeout` would be enough?

Reply:
For correct debounce behaviour, it will cancel the previous timer everytime when the function is called. So we need to change the code for debounce behaviour or use lodash debounce or maybe `useDebounce` hook.

I think we want something similar to the usage of React `useDeferredValue`, refer to [this section in react doc](https://reactjs.org/docs/concurrent-mode-patterns.html#deferring-a-value). We may also need to expose a debounced `setUpdatedValue` function (not so sure abt this)

Reply:
I might be missing something here, but isn't `clearTimeout` doing exactly that? Also about exposing the function, I don't think we should. This hooks supposed to expose only the value in my opinion.

Reply:
Set myself a reminder to comeback to this in the weekend. Will get it shipped in one form or another tomorrow @maciekgrzybek @qiweiii . Thanks for your hardwork!

Reply:
What do you think of this? @maciekgrzybek @qiweiii ? Just checking if this solves the problem and to look for the best way of doing this. 👍🏽 

```jsx

function useDebouncedValue<T = unknown>(value: T, timeout: number) {
  const [updatedValue, setUpdatedValue] = useState<T | null>(null);
  const debouncedSetUpdatedValue = useDebounce(setUpdatedValue);
  useEffect(() => {
       debouncedSetUpdatedValue(updatedValue);
  }, [updatedValue])
 // No need to add `debouncedSetUpdatedValue ` to dependencies as it is a ref.current.

  return updatedValue;

}


```

Reply:
I don't want to overcomplicate it but at the same time I think `setUpdatedValue` might be useful here.

Reply:
This is not working unfortunately, the tests are failing. I tried something similar before - same result.

Reply:
I also tried something similar but tests are failing. Found this [issue](https://github.com/facebook/jest/issues/3465) about jest.runAllTimers may be related